File: src/main/java/io/ercole/model/Alert.java
// Copyright (c) 2019 Sorint.lab S.p.A.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package io.ercole.model;

import java.util.Date;

import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Lob;

import org.hibernate.annotations.Type;

/**
 * The Class Alert.
 */
@Entity
public class Alert {


	/** The id. */
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	/** The hostname. */
	private String hostname;

	/** The code. */
	@Enumerated(EnumType.STRING)
	private AlertCode code;

	/** The severity. */
	@Enumerated(EnumType.STRING)
	private AlertSeverity severity;
	
	/** The status. */
	@Enumerated(EnumType.STRING)
	private AlertStatus status;

	/** The description. */
	@Lob
	@Type(type = "org.hibernate.type.TextType")
	private String description;
	
	/** The date. */
	private Date date;
	
	/**
	 * Instantiates a new alert.
	 */
	public Alert() {
		
	}
	/**
	 * Instantiates a new alert.
	 *
	 * @param name the hostname
	 * @param cod the code
	 * @param descr the description
	 * @param sever the severity
	 */
	public Alert(final String name, final AlertCode cod, final String descr, 
			final AlertSeverity sever) {
		this.hostname = name;
		this.code = cod;
		this.description = descr;
		this.severity = sever;
		date = new Date();
		status = AlertStatus.NEW;
	}
	
	
	



	/**
	 * Gets the id.
	 *
	 * @return the id
	 */
	public Long getId() {
		return id;
	}

	/**
	 * Sets the id.
	 *
	 * @param id the new id
	 */
	public void setId(final Long id) {
		this.id = id;
	}

	/**
	 * Gets the hostname.
	 *
	 * @return the hostname
	 */
	public String getHostname() {
		return hostname;
	}

	/**
	 * Sets the hostname.
	 *
	 * @param hostname the new hostname
	 */
	public void setHostname(final String hostname) {
		this.hostname = hostname;
	}

	/**
	 * Gets the code.
	 *
	 * @return the code
	 */
	public AlertCode getCode() {
		return code;
	}

	/**
	 * Sets the code.
	 *
	 * @param code the new code
	 */
	public void setCode(final AlertCode code) {
		this.code = code;
	}

	/**
	 * Gets the severity.
	 *
	 * @return the severity
	 */
	public AlertSeverity getSeverity() {
		return severity;
	}

	/**
	 * Sets the severity.
	 *
	 * @param severity the new severity
	 */
	public void setSeverity(final AlertSeverity severity) {
		this.severity = severity;
	}

	/**
	 * Gets the description.
	 *
	 * @return the description
	 */
	public String getDescription() {
		return description;
	}

	/**
	 * Sets the description.
	 *
	 * @param description the new description
	 */
	public void setDescription(final String description) {
		this.description = description;
	}

	/**
	 * Gets the date.
	 *
	 * @return the date
	 */
	public Date getDate() {
		return date;
	}

	/**
	 * Sets the date.
	 *
	 * @param date the new date
	 */
	public void setDate(final Date date) {
		this.date = date;
	}
	
	/**
	 * Get the status.
	 * @return the alerts's status
	 */
	public AlertStatus getStatus() {
		return status;
	}
	
	/**
	 * Set the status.
	 * @param status the alert status
	 */
	public void setStatus(final AlertStatus status) {
		this.status = status;
	}
	
	
	
	

}

File: model/alert.go
// Copyright (c) 2020 Sorint.lab S.p.A.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package model

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Alert holds informations about a alert
type Alert struct {
	ID                      primitive.ObjectID     `json:"id" bson:"_id"`
	AlertCategory           string                 `json:"alertCategory" bson:"alertCategory"`
	AlertAffectedTechnology *string                `json:"alertAffectedTechnology" bson:"alertAffectedTechnology"`
	AlertCode               string                 `json:"alertCode" bson:"alertCode"`
	AlertSeverity           string                 `json:"alertSeverity" bson:"alertSeverity"`
	AlertStatus             string                 `json:"alertStatus" bson:"alertStatus"`
	Description             string                 `json:"description" bson:"description"`
	Date                    time.Time              `json:"date" bson:"date"`
	OtherInfo               map[string]interface{} `json:"otherInfo" bson:"otherInfo"`
}

const (
	AlertCategoryEngine  string = "ENGINE"
	AlertCategoryAgent   string = "AGENT"
	AlertCategoryLicense string = "LICENSE"
)

func getAlertCategories() []string {
	return []string{AlertCategoryEngine, AlertCategoryAgent, AlertCategoryLicense}
}

const (
	// ENGINE

	AlertCodeNewServer               string = "NEW_SERVER"
	AlertCodeUnlistedRunningDatabase string = "UNLISTED_RUNNING_DATABASE"
	AlertCodeMissingPrimaryDatabase  string = "MISSING_PRIMARY_DATABASE"

	// AGENT

	AlertCodeNoData string = "NO_DATA"

	// LICENSE

	AlertCodeNewDatabase string = "NEW_DATABASE"
	AlertCodeNewLicense  string = "NEW_LICENSE"
	AlertCodeNewOption   string = "NEW_OPTION"
)

func getAlertCodes() []string {
	return []string{
		AlertCodeNewServer, AlertCodeUnlistedRunningDatabase, AlertCodeMissingPrimaryDatabase,
		AlertCodeNoData,
		AlertCodeNewDatabase, AlertCodeNewLicense, AlertCodeNewOption,
	}
}

const (
	AlertSeverityInfo     string = "INFO"
	AlertSeverityWarning  string = "WARNING"
	AlertSeverityCritical string = "CRITICAL"
)

func getAlertSeverities() []string {
	return []string{AlertSeverityInfo, AlertSeverityWarning, AlertSeverityCritical}
}

// Alert status
const (
	// New contains string "NEW"
	AlertStatusNew string = "NEW"
	// Ack contains string "ACK"
	AlertStatusAck string = "ACK"
)

func getAlertStatuses() []string {
	return []string{AlertStatusNew, AlertStatusAck}
}

func (alert Alert) IsValid() bool {
	fields := make(map[string][]string)
	fields[alert.AlertCategory] = getAlertCategories()
	fields[alert.AlertCode] = getAlertCodes()
	fields[alert.AlertSeverity] = getAlertSeverities()
	fields[alert.AlertStatus] = getAlertStatuses()

fields:
	for thisValue, allValidValues := range fields {
		for _, validValue := range allValidValues {
			if thisValue == validValue {
				continue fields
			}
		}

		return false
	}

	return true
}
